[ao][sparsity] Bug Fix: Retain mask and config while replacing data in data sparsifier

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82131
* #82130
* __->__ #82129

Bug: The config and mask were being recreated while replacing data on the data sparsifier.

Fix: Introduced an argument `reuse_mask` which when set `True` uses the old mask. If new config is not
specified, the data sparsifier by default uses the old config with the new data.
Also, added unit tests to check this bug.

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Commit:
[ao][sparsity] Bug Fix: Retain mask and config while replacing data in data sparsifier

Bug: The config and mask were being recreated while replacing data on the data sparsifier.

Fix: Introduced an argument `use_old_mask` which when set `True` uses the old mask. If new config is not
specified, the data sparsifier by default uses the old config with the new data.
Also, added unit tests to check this bug.

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Bug Fix: Retain mask and config while replacing data in data sparsifier"


Bug: The config and mask were being recreated while replacing data on the data sparsifier.

Fix: Introduced an argument `use_old_mask` which when set `True` uses the old mask. If new config is not
specified, the data sparsifier by default uses the old config with the new data.
Also, added unit tests to check this bug.

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Bug Fix: Retain mask and config while replacing data in data sparsifier"


Bug: The config and mask were being recreated while replacing data on the data sparsifier.

Fix: Introduced an argument `use_old_mask` which when set `True` uses the old mask. If new config is not
specified, the data sparsifier by default uses the old config with the new data.
Also, added unit tests to check this bug.

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

z-a-f commented on torch/ao/sparsity/_experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -4,10 +4,14 @@
+import sys
+import warnings
+
+if not sys.warnoptions:
+    warnings.simplefilter("once")
```
nit: Can you add a comment here why the warnings are filtered?

z-a-f commented on torch/ao/sparsity/_experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -74,8 +78,15 @@ def _extract_weight(self, data):
 
-    def add_data(self, name: str, data, **config):
-        r""" Configures and parametrizes the internal container model with name and data
+    def add_data(self, name: str, data, use_old_mask=True, **config):
```
Can we rename the `use_old_mask` into `reuse_mask` (or something like that)? imho, it's avoids the questions, such as "what do you mean by old?"

Commit:
Update on "[ao][sparsity] Bug Fix: Retain mask and config while replacing data in data sparsifier"


Bug: The config and mask were being recreated while replacing data on the data sparsifier.

Fix: Introduced an argument `use_old_mask` which when set `True` uses the old mask. If new config is not
specified, the data sparsifier by default uses the old config with the new data.
Also, added unit tests to check this bug.

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Bug Fix: Retain mask and config while replacing data in data sparsifier"


Bug: The config and mask were being recreated while replacing data on the data sparsifier.

Fix: Introduced an argument `reuse_mask` which when set `True` uses the old mask. If new config is not
specified, the data sparsifier by default uses the old config with the new data.
Also, added unit tests to check this bug.

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

macandro96 commented:
@pytorchbot merge -g
